test/examples/ucp_client_server.c

```diff
@@ -0,0 +1,450 @@
+     *                                        is temporarily required since the
+     *                                        endpoint will be closed with
+     *                                        UCP_EP_CLOSE_MODE_FORCE which
+     *                                        requires this mode.
+     *                                        Once UCP_EP_CLOSE_MODE_FORCE is
+     *                                        removed, the error handling mode
+     *                                        will be removed.
+     */
+    ep_params.field_mask       = UCP_EP_PARAM_FIELD_FLAGS     |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&connect_addr;
+    ep_params.sockaddr.addrlen = sizeof(connect_addr);
+
+    status = ucp_ep_create(ucp_worker, &ep_params, client_ep);
+    if (status != UCS_OK) {
+        fprintf(stderr, "failed to connect to %s\n", ip);
+    }
+
+    return status;
+}
+
+/**
+ * Verify the received message on the server side and print the result.
+ */
+static int verify_result(int is_server, char *recv_message)
+{
+    int ret;
+
+    if (is_server) {
+        if (!strcmp(recv_message, test_message)) {
+            printf("\n\n----- UCP TEST SUCCESS -------\n\n");
+            printf("%s", recv_message);
+            printf("\n\n------------------------------\n\n");
+
+            ret = 0;
+        } else {
+            printf("\n\n----- UCP TEST FAILURE -------\n\n");
+            printf("Client sent message:     %s\nServer received message: %s\n",
+                   test_message, recv_message);
+            printf("\n\n------------------------------\n\n");
+
+            ret = -1;
+        }
+    } else {
+        printf("\n\n-----------------------------------------\n\n");
+        printf("Client sent message: \n%s.\nlength: %ld\n",
+               test_message, TEST_STRING_LEN);
+        printf("\n-----------------------------------------\n\n");
+        ret = 0;
+    }
+
+    return ret;
+}
+
+/**
+ * Send and receive a message using the Stream API.
+ * The client sends a message to the server and waits until the send it completed.
+ * The server receives a message from the client and waits for its completion.
+ */
+static int send_recv_stream(ucp_worker_h ucp_worker, ucp_ep_h ep, int is_server)
+{
+    char recv_message[TEST_STRING_LEN]= "";
+    test_req_t *request;
+    size_t length;
+    int ret;
+
+    if (!is_server) {
+        /* Client sends a message to the server using the stream API */
+        request = ucp_stream_send_nb(ep, test_message, 1,
+                                     ucp_dt_make_contig(TEST_STRING_LEN),
+                                     stream_send_cb, 0);
+        if (UCS_PTR_IS_ERR(request)) {
```

Comment:
can add a convenience function request_wait() which will check wait for completion and release the request